# Portfolio Project: Online Retail Exploratory Data Analysis with Python

## Overview

In this project, you will step into the shoes of an entry-level data analyst at an online retail company, helping interpret real-world data to help make a key business decision.

Source files can be found here: https://www.coursera.org/learn/perform-exploratory-data-analysis-on-retail-data-with-python/home/week/1

## Case Study
In this project, you will be working with transactional data from an online retail store. The dataset contains information about customer purchases, including product details, quantities, prices, and timestamps. Your task is to explore and analyze this dataset to gain insights into the store's sales trends, customer behavior, and popular products. 

By conducting exploratory data analysis, you will identify patterns, outliers, and correlations in the data, allowing you to make data-driven decisions and recommendations to optimize the store's operations and improve customer satisfaction. Through visualizations and statistical analysis, you will uncover key trends, such as the busiest sales months, best-selling products, and the store's most valuable customers. Ultimately, this project aims to provide actionable insights that can drive strategic business decisions and enhance the store's overall performance in the competitive online retail market.

## Project Objectives
1. Describe data to answer key questions to uncover insights
2. Gain valuable insights that will help improve online retail performance
3. Provide analytic insights and data-driven recommendations

## Dataset

The dataset you will be working with is the "Online Retail" dataset. It contains transactional data of an online retail store from 2010 to 2011. The dataset is available as a .xlsx file named `Online Retail.xlsx`. This data file is already included in the Coursera Jupyter Notebook environment, however if you are working off-platform it can also be downloaded [here](https://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx).

The dataset contains the following columns:

- InvoiceNo: Invoice number of the transaction
- StockCode: Unique code of the product
- Description: Description of the product
- Quantity: Quantity of the product in the transaction
- InvoiceDate: Date and time of the transaction
- UnitPrice: Unit price of the product
- CustomerID: Unique identifier of the customer
- Country: Country where the transaction occurred

## Tasks

You may explore this dataset in any way you would like - however if you'd like some help getting started, here are a few ideas:

1. Load the dataset into a Pandas DataFrame and display the first few rows to get an overview of the data.
2. Perform data cleaning by handling missing values, if any, and removing any redundant or unnecessary columns.
3. Explore the basic statistics of the dataset, including measures of central tendency and dispersion.
4. Perform data visualization to gain insights into the dataset. Generate appropriate plots, such as histograms, scatter plots, or bar plots, to visualize different aspects of the data.
5. Analyze the sales trends over time. Identify the busiest months and days of the week in terms of sales.
6. Explore the top-selling products and countries based on the quantity sold.
7. Identify any outliers or anomalies in the dataset and discuss their potential impact on the analysis.
8. Draw conclusions and summarize your findings from the exploratory data analysis.

# Task 1: Load the Data

In [3]:
import pandas as pd
pd.set_option('max_colwidth', 400)

In [4]:
%%timeit
retail_data = pd.read_excel('Online Retail.xlsx', sheet_name='Online Retail')

In [76]:
retail_data.head(500)

InvoiceNo StockCode                          Description  Quantity  \
0      536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1      536365     71053                  WHITE METAL LANTERN         6   
2      536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3      536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4      536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
..        ...       ...                                  ...       ...   
495    536409     20669                RED HEART LUGGAGE TAG         1   
496    536409    90129F           RED GLASS TASSLE BAG CHARM         1   
497    536409    90210B         CLEAR ACRYLIC FACETED BANGLE         1   
498    536409    90199C      5 STRAND GLASS NECKLACE CRYSTAL         1   
499    536409     21955    DOORMAT UNION JACK GUNS AND ROSES         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
0   2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1   2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2   2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3   2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4   2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
..                  ...        ...         ...             ...  
495 2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
496 2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
497 2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
498 2010-12-01 11:45:00       6.35     17908.0  United Kingdom  
499 2010-12-01 11:45:00       7.95     17908.0  United Kingdom  

[500 rows x 8 columns]

In [77]:
# display(retail_data.describe())
# display(retail_data.info())
display(retail_data.isnull().sum())
# display(retail_data.nunique())

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

| Column      | Unique Values | Null Values |
|:-------------:|:---------------:|:-------------:|
| InvoiceNo   | 25900         | 0           |
| StockCode   | 4070          | 0           |
| Description | 4223          | 1454        |
| Quantity    | 722           | 0           |
| InvoiceDate | 23260         | 0           |
| UnitPrice   | 1630          | 0           |
| CustomerID  | 4372          | 135080      |
| Country     | 38            | 0           |

This dataset contains 541,909 rows.

It's alarming that 135,080 (25%) of those rows contain null `CustomerID` values. This is a piece of important information to provide back to the store owner.  Additionally, 1,454 rows are missing a `Description` but, while that's a problem, the corresponding StockCode doesn't seem to be missing.  `Description` would be helpful to have from a readability standpoint but probably isn't needed for EDA.

I am choosing to keep the row with missing `CustomerID` rows for now because I may be able to do analysis on the data set (E.g., product analysis, time analysis, geospacial analysis) that won't require the data.

# Task 2: Clean the data


To Do Ideas:
1. Identify blank `Descriptions` and populate the correct value based on StockCode (Code of the Product), if possible.
2. Identify blank `CustomerID` and populate the correct valuea based on InvoiceNo, since multiple customers shouldn't be on the same Invoice?
   

## Fix missing Description values

Possible process:

- Create a list of `StockCode` and the most frequent corresponding `Description`. E.g. for StockCode "10080", there are three unique `Description` values: "check" (1), "GROOVY CACTUS INFLATABLE" (22), and "blank" (1). The most common `Description` is likely the valid value.
- Replace the existing value with the most-popular value
- I don't think it'll be easier to filter `StockCode` that have more than one unique `Description` but maybe.

### Count unique Descriptions per StockCode
I begin by counting the unique `Description` values per `StockCode`. Counts > 1 indiciate that the corresponding `StockCode` has more than one `Description` which, based on the description of the Dataset's columns, this shouldn't be possible.

In [78]:
unique_description_counts = retail_data.groupby('StockCode')['Description'].nunique().reset_index()
display(unique_description_counts[unique_description_counts['Description'] > 1])

StockCode  Description
1            10080            2
4            10133            2
12           16008            2
21           16045            2
50           20622            2
...            ...          ...
3974        90195A            2
4008        90210D            2
4043      DCGS0003            2
4050      DCGS0069            2
4065  gift_0001_20            2

[650 rows x 2 columns]

### Create a dictionary of Key:Value pairs using StockCode and most-popular Descriptions

The `most_common` function takes a single argument. If the Series is entirely composed of missing values, the function returns a None value. Otherwise, it returns the mode of each value in the `StockCode` column.

In [91]:
# Define a function to find the most common value or return None if the Series is empty.
def most_common(series):
    if series.dropna().empty:
        return None
    return series.mode().iloc[0]

# Creates a Lookup of all StockCode and hopefully-correct Descriptions. Group by 'StockCode' and apply the most_common function to 'Description'
stockcode_description_mode = retail_data.groupby('StockCode')['Description'].apply(most_common).reset_index()
display(stockcode_description_mode)

StockCode                         Description
0            10002         INFLATABLE POLITICAL GLOBE 
1            10080            GROOVY CACTUS INFLATABLE
2            10120                        DOGGY RUBBER
3            10125             MINI FUNKY DESIGN TAPES
4            10133        COLOURING PENCILS BROWN TUBE
...            ...                                 ...
4065  gift_0001_20  Dotcomgiftshop Gift Voucher £20.00
4066  gift_0001_30  Dotcomgiftshop Gift Voucher £30.00
4067  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00
4068  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00
4069             m                              Manual

[4070 rows x 2 columns]

The `stockcode_description_mode` DataFrame above contains 4070 rows, which matches the number UniqueValues identified in Task 1 for `StockCodes`. This suggests the code, so far, is correct.

I convert `stockcode_description_mode` to a dictionary `stockcode_description_dict` because I was getting errors in a step that used the `.replace()` and `.map()` functions.

In [ ]:
# Convert DataFrame to dictionary
stockcode_description_dict = dict(zip(stockcode_description_mode['StockCode'], stockcode_description_mode['Description']))
# display(stockcode_description_dict)

I was not able to get the `.replace()` and `.map()` functions to "find and replace" the `StockCode` and `Description` values. As an alternative solution, I use the `.merge()` method to accomplish the same results. I clean up the results by dropping the original `Description` column (now `Description_x`), naming the new `Description_y` to `Description`, and reordering the columns.  

In [80]:
# Merge dictionary to retail_data
retail_data_new = pd.merge(retail_data, stockcode_description_mode, on='StockCode').drop(columns='Description_x').rename(columns={'Description_y':'Description'})

# Fix Column Order and Rename
retail_data_new = retail_data_new[['InvoiceNo', 'StockCode','Description', 'Quantity','InvoiceDate','UnitPrice','CustomerID','Country']]
display(retail_data_new)

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536373    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
2         536375    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
3         536390    85123A   WHITE HANGING HEART T-LIGHT HOLDER        64   
4         536394    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32   
...          ...       ...                                  ...       ...   
541904    579512     23617  SET 10 CARDS SWIRLY XMAS TREE 17104         2   
541905    580691    90214U            LETTER "U" BLING KEY RING        12   
541906    581238    47591b             SCOTTIES CHILDRENS APRON         1   
541907    581483     23843          PAPER CRAFT , LITTLE BIRDIE     80995   
541908   C581484     23843          PAPER CRAFT , LITTLE BIRDIE    -80995   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 09:02:00       2.55     17850.0  United Kingdom  
2      2010-12-01 09:32:00       2.55     17850.0  United Kingdom  
3      2010-12-01 10:19:00       2.55     17511.0  United Kingdom  
4      2010-12-01 10:39:00       2.55     13408.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-11-29 16:47:00       2.91         NaN  United Kingdom  
541905 2011-12-05 15:48:00       0.29     13790.0  United Kingdom  
541906 2011-12-08 10:53:00       4.13         NaN  United Kingdom  
541907 2011-12-09 09:15:00       2.08     16446.0  United Kingdom  
541908 2011-12-09 09:27:00       2.08     16446.0  United Kingdom  

[541909 rows x 8 columns]

### Reviewing the fix

In [94]:
filt = retail_data_new['StockCode'] == 10080
retail_data_new.loc[filt]

InvoiceNo StockCode               Description  Quantity  \
447801    545060     10080  GROOVY CACTUS INFLATABLE         2   
447802    557568     10080  GROOVY CACTUS INFLATABLE        48   
447803    558614     10080  GROOVY CACTUS INFLATABLE        12   
447804    558911     10080  GROOVY CACTUS INFLATABLE        12   
447805    561271     10080  GROOVY CACTUS INFLATABLE       170   
447806    561322     10080  GROOVY CACTUS INFLATABLE        12   
447807    561925     10080  GROOVY CACTUS INFLATABLE        12   
447808    562044     10080  GROOVY CACTUS INFLATABLE        24   
447809    562127     10080  GROOVY CACTUS INFLATABLE        12   
447810    564065     10080  GROOVY CACTUS INFLATABLE        12   
447811    565618     10080  GROOVY CACTUS INFLATABLE        12   
447812    565920     10080  GROOVY CACTUS INFLATABLE        24   
447813    567200     10080  GROOVY CACTUS INFLATABLE        12   
447814    568381     10080  GROOVY CACTUS INFLATABLE        12   
447815    570458     10080  GROOVY CACTUS INFLATABLE         1   
447816    571046     10080  GROOVY CACTUS INFLATABLE         3   
447817    572569     10080  GROOVY CACTUS INFLATABLE         2   
447818    574036     10080  GROOVY CACTUS INFLATABLE        24   
447819    575518     10080  GROOVY CACTUS INFLATABLE        22   
447820    575908     10080  GROOVY CACTUS INFLATABLE        24   
447821    576095     10080  GROOVY CACTUS INFLATABLE        12   
447822    577300     10080  GROOVY CACTUS INFLATABLE         4   
447823    577773     10080  GROOVY CACTUS INFLATABLE         1   
447824    577801     10080  GROOVY CACTUS INFLATABLE        26   

               InvoiceDate  UnitPrice  CustomerID         Country  
447801 2011-02-27 13:47:00       0.85     15547.0  United Kingdom  
447802 2011-06-21 10:40:00       0.39     17629.0  United Kingdom  
447803 2011-06-30 15:56:00       0.39         NaN  United Kingdom  
447804 2011-07-05 09:23:00       0.39     14189.0  United Kingdom  
447805 2011-07-26 12:18:00       0.00         NaN  United Kingdom  
447806 2011-07-26 13:03:00       0.39     15203.0  United Kingdom  
447807 2011-08-01 12:28:00       0.39     16551.0  United Kingdom  
447808 2011-08-02 10:24:00       0.39     16607.0  United Kingdom  
447809 2011-08-03 08:21:00       0.39     13717.0  United Kingdom  
447810 2011-08-22 15:05:00       0.39     15150.0  United Kingdom  
447811 2011-09-05 15:40:00       0.39     15974.0  United Kingdom  
447812 2011-09-08 08:56:00       0.39     15298.0  United Kingdom  
447813 2011-09-19 10:15:00       0.39     16460.0  United Kingdom  
447814 2011-09-27 09:24:00       0.39     14321.0  United Kingdom  
447815 2011-10-10 15:39:00       0.39     14535.0  United Kingdom  
447816 2011-10-13 13:35:00       0.39     18096.0  United Kingdom  
447817 2011-10-25 10:32:00       0.39     17870.0  United Kingdom  
447818 2011-11-02 12:46:00       0.39     16607.0  United Kingdom  
447819 2011-11-10 10:53:00       0.00         NaN  United Kingdom  
447820 2011-11-11 15:54:00       0.39     13091.0  United Kingdom  
447821 2011-11-14 09:13:00       0.39     14904.0  United Kingdom  
447822 2011-11-18 12:53:00       0.39     16924.0  United Kingdom  
447823 2011-11-21 15:57:00       0.39     16712.0  United Kingdom  
447824 2011-11-21 17:04:00       0.39     17629.0  United Kingdom

`StockCode` 10080 had two "extra" `Description` values: "check" and a null value. Now, the "GROOVY CACTUS INFLATABLE" value is present for all 10080 rows!

When I re-run the same summary funcations on the new `retail_data_new` dataset, the `StockCode` numbers are significantly better.

In [96]:
# display(retail_data_new.describe())
# display(retail_data_new.info())
display(retail_data_new.isnull().sum())
# display(retail_data_new.nunique())

InvoiceNo           0
StockCode           0
Description       112
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [100]:
filt_stockcode_description_mode = stockcode_description_mode[stockcode_description_mode['Description'].isna()]
filt_results = filt_stockcode_description_mode[['StockCode', 'Description']]
print(filt_results)

      StockCode Description
5         10134        None
26        16053        None
80        20689        None
111       20738        None
159       20825        None
...         ...         ...
4046   DCGS0057        None
4047  DCGS0066P        None
4052   DCGS0071        None
4053   DCGS0072        None
4055   DCGS0074        None

[112 rows x 2 columns]


I wanted to investigate the 112 `StockCode`s that were still blank. It turns out the corresponding `Description` values are all blank. Coincidentally, the row's `CustomerID` values are also blank. It would be my recommendation that these rows be dropped from the entire dataset.

## Fix missing CustomerID values 

This cannot be done the same way I fixed the Descriptions. When an InvoiceNo is missing a CustomerID, it's gone. For a given InvoiceID that has at least one missing CustomerID, there aren't other rows that have a CustomerID that have the same InvoiceNo.

The current dataset does not have enough information to execute a creative solution to impute missing data.

# Task 3: Basic Statistics

## Central Tendency
### Mean
**Mean (Average)** is the average value for a column. It is calculated as the sum of all values in the column divided by the number of data points.
- Mean of Quantity: mean_quantity = sum(Quantity) / count(Quantity)
- Mean of UnitPrice: mean_unit_price = sum(UnitPrice) / count(UnitPrice)

### Median
**Median** is the middle value in the sorted data. It's a measure of central tendency that is less affected by outliers.
- Median of Quantity: Sort the Quantity column and select the middle value.
- Median of UnitPrice: Sort the UnitPrice column and select the middle value.

### Mode
**Mode** is the most frequently occurring value in a column.
- Mode of Quantity: The value(s) that occur most frequently in the Quantity column.
- Mode of UnitPrice: The value(s) that occur most frequently in the UnitPrice column.

## Dispersion

### Range

The **range** gives you an idea of the spread of the data. It is calculated as the difference between the maximum and minimum values.
- Range of Quantity: range_quantity = max(Quantity) - min(Quantity)
- Range of UnitPrice: range_unit_price = max(UnitPrice) - min(UnitPrice)

### Variance

**Variance** measures the average squared difference of each data point from the mean. It quantifies how much the data points deviate from the mean.
- Variance of Quantity: `variance_quantity = sum((Quantity - mean_quantity)^2) / (count(Quantity) - 1)`
- Variance of UnitPrice: `variance_unit_price = sum((UnitPrice - mean_unit_price)^2) / (count(UnitPrice) - 1)`

### Standard Deviation

**Standard deviation** is the square root of the variance. It measures the average deviation from the mean and is often easier to interpret than variance.
- Standard deviation of Quantity: std_dev_quantity = sqrt(variance_quantity)
- Standard deviation of UnitPrice: std_dev_unit_price = sqrt(variance_unit_price)

### Interquartile Range
**Interquartile Range (IQR)** is the range between the first quartile (Q1) and the third quartile (Q3) of the data. It helps to understand the middle 50% of the data.
- Calculate Q1 and Q3, then IQR = Q3 - Q1 for Quantity and UnitPrice.